### ISSUE TITLE
Support new Docker Swarm Mode (Docker 17+)

### ISSUE THREAD

**Issue by scnerd** [jupyterhub/dockerspawner#215]
It's official, [SwarmSpawner is deprecated.](https://github.com/cassinyio/SwarmSpawner/issues/29) After several days of copying every example configuration I can find, I can't get the latest SwarmSpawner to work on Docker 17+ (specifically 17.12.0-ce).

Any chance spawning Docker Services into a new-style Docker Swarm will get supported any time soon? This repository seems like the perfect place to implement such a feature (seeing as old-style swarms were already supported), now that there's no other alternative.

**Comment by scnerd** [comment]
I'm currently working on implementing this in [scnerd:swarmspawner](https://github.com/scnerd/dockerspawner/tree/swarmspawner) and it looks like there's some older, possibly stale work in [yacchin1205:with-docker-services-api](https://github.com/yacchin1205/dockerspawner/tree/with-docker-services-api). Neither approach looks nearly mature, but I'm hoping to get my branch robust enough to propose a PR.

**Comment by co60ca** [comment]
FYI I patched version 3906f4bebc92b383c73fb8d06c58a7c57003939a

with
```
diff --git a/dockerspawner/swarmspawner.py b/dockerspawner/swarmspawner.py
index 501047a..f853f0c 100644
--- a/dockerspawner/swarmspawner.py
+++ b/dockerspawner/swarmspawner.py
@@ -105,7 +105,7 @@ class SwarmSpawner(DockerSpawner):
                     source=host_loc,
                     type="bind",
                     read_only=vol["mode"] == "ro",
-                    driver_config=driver,
+                #    driver_config=driver,
                 )
                 for host_loc, vol in self.volume_binds.items()
             ]
@@ -126,7 +126,7 @@ class SwarmSpawner(DockerSpawner):
             "Service %s status: %s", self.service_id[:7], pformat(service_state)
         )

-        if service_state["State"] in {"running", "pending"}:
+        if service_state["State"] in {"running", "pending", "starting"}:
             return None

         else:
```
and it works for me. You'll note that this isn't a proper fix but might help someone anyway.

**Comment by minrk** [comment]
Hi! I’m going through and cleaning up old/stale issues on this repo.

SwarmSpawner has received several updates related to handling service states (e.g. #296), so I think this is probably fixed. Feel free to open a new Issue if you are still having this trouble.
